linkerd/retry/src/lib.rs

```diff
@@ -93,35 +105,42 @@ where
 
 // === impl Retry ===
 
-impl<R, P, Req, S> Proxy<Req, S> for Retry<R, P>
+impl<R, P, B, S> Proxy<http::Request<B>, S> for Retry<R, P>
 where
-    R: tower::retry::Policy<Req, P::Response, Error> + Clone,
-    P: Proxy<Req, S> + Clone,
+    R: tower::retry::Policy<http::Request<ReplayBody<B>>, P::Response, Error> + Clone,
+    P: Proxy<http::Request<replay::MaybeReplay<B>>, S> + Clone,
     S: tower::Service<P::Request> + Clone,
     S::Error: Into<Error>,
+    B: http_body::Body + Unpin,
 {
     type Request = P::Request;
     type Response = P::Response;
     type Error = Error;
-    type Future = ResponseFuture<R, P, S, Req>;
+    type Future = ResponseFuture<R, P, S, B>;
 
-    fn proxy(&self, svc: &mut S, req: Req) -> Self::Future {
+    fn proxy(&self, svc: &mut S, req: http::Request<B>) -> Self::Future {
         trace!(retryable = %self.policy.is_some());
 
         if let Some(policy) = self.policy.as_ref() {
-            let inner = self.inner.clone().wrap_service(svc.clone());
+            let inner = self.inner.clone().wrap_service(svc.clone()).map_request(
+                (|req: http::Request<ReplayBody<B>>| req.map(replay::MaybeReplay::Replay))
+                    as fn(_) -> _,
+            );
             let retry = tower::retry::Retry::new(policy.clone(), inner);
-            return ResponseFuture::Retry(retry.oneshot(req));
+            return ResponseFuture::Retry(retry.oneshot(req.map(ReplayBody::new)));
         }
 
-        ResponseFuture::Disabled(self.inner.proxy(svc, req))
+        ResponseFuture::Disabled(
+            self.inner
+                .proxy(svc, req.map(replay::MaybeReplay::NoReplay)),
```

Comment:
And if this really just an `EitherBody`, can we just use `BoxBody` instead? Or do the `Data` types not line up?